relay-server/src/utils/metrics_rate_limits.rs

```diff
@@ -130,112 +131,212 @@ pub enum ExtractionMode {
     Duration,
 }
 
-/// Return value of [`count_metric_bucket`], containing the extracted
-/// count of transactions and wether they have associated profiles.
-#[derive(Debug, Clone, Copy)]
-struct TransactionCount {
-    /// Number of transactions.
-    pub count: usize,
-    /// Whether the transactions have associated profiles.
-    pub has_profile: bool,
+/// The return value of [`count_metric_bucket`].
+///
+/// Contains the count of total transactions or spans that went into this bucket.
+#[derive(Debug, Default, Clone)]
+pub enum BucketSummary {
+    Transactions {
+        count: usize,
+        has_profile: bool,
+    },
+    Spans(usize),
+    #[default]
+    None,
+}
+
+impl BucketSummary {
+    fn to_counts(&self) -> TotalEntityCounts {
+        match *self {
+            BucketSummary::Transactions { count, has_profile } => TotalEntityCounts {
+                transactions: Some(count),
+                spans: None,
+                profiles: if has_profile { count } else { 0 },
+            },
+            BucketSummary::Spans(count) => TotalEntityCounts {
+                transactions: None,
+                spans: Some(count),
+                profiles: 0,
+            },
+            BucketSummary::None => TotalEntityCounts::default(),
+        }
+    }
+}
+
+#[derive(Debug)]
+struct SummarizedBucket {
+    bucket: Bucket,
+    summary: BucketSummary,
+}
+
+#[derive(Debug, Default, Clone)]
+struct TotalEntityCounts {
+    transactions: Option<usize>,
+    spans: Option<usize>,
+    profiles: usize,
+}
+
+impl std::ops::Add for TotalEntityCounts {
+    type Output = Self;
+
+    fn add(self, rhs: TotalEntityCounts) -> Self::Output {
+        Self {
+            transactions: add_some(self.transactions, rhs.transactions),
+            spans: add_some(self.spans, rhs.spans),
+            profiles: self.profiles + rhs.profiles,
+        }
+    }
+}
+
+fn add_some<T>(a: Option<T>, b: Option<T>) -> Option<T>
+where
+    T: std::ops::Add<Output = T>,
+{
+    match (a, b) {
+        (None, None) => None,
+        (None, Some(c)) | (Some(c), None) => Some(c),
+        (Some(a), Some(b)) => Some(a + b),
+    }
 }
 
 impl<Q: AsRef<Vec<Quota>>> MetricsLimiter<Q> {
     /// Create a new limiter instance.
     ///
-    /// Returns Ok if `metrics` contain transaction metrics, `metrics` otherwise.
+    /// Returns Ok if `metrics` contain relevant metrics, `metrics` otherwise.
     pub fn create(
         buckets: Vec<Bucket>,
         quotas: Q,
         scoping: Scoping,
         mode: ExtractionMode,
     ) -> Result<Self, Vec<Bucket>> {
-        let counts: Vec<_> = buckets
-            .iter()
-            .map(|metric| count_metric_bucket(BucketView::new(metric), mode))
+        let buckets: Vec<_> = buckets
+            .into_iter()
```

Comment:
Simplifies code, but this results in an additional allocation, which is probably fine, just wanna bring it up.

Reply:
Does it? `into_iter()` takes ownership of the existing vector AFAIK.

Reply:
`into_iter()` takes ownership, but each element of the vector changes size -> needs an allocation, later we extract the buckets again -> another allocation. Instead of just having 1 allocation for the extra data.